File: app/src/test/java/cz/civilizacehra/cipherbreaker/NumberAnalysisTest.kt
package cz.civilizacehra.cipherbreaker

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class NumberAnalysisTest {

    private fun factors(vararg factors: Long): List<ULong> {
        return factors.map { it.toULong() }
    }

    @Test
    fun numbersBelowTwoHaveNoFactors() {
        assertEquals(factors(), NumberAnalysis.factorNumber(0u))
        assertEquals(factors(), NumberAnalysis.factorNumber(1u))
    }

    @Test
    fun primesAreTheirOnlyFactor() {
        assertEquals(factors(2), NumberAnalysis.factorNumber(2u))
        assertEquals(factors(3), NumberAnalysis.factorNumber(3u))
        assertEquals(factors(97), NumberAnalysis.factorNumber(97u))
        assertEquals(factors(1000000007), NumberAnalysis.factorNumber(1000000007u))
    }

    @Test
    fun factorsAreSortedAndRepeated() {
        assertEquals(factors(2, 2, 2, 3, 3, 5), NumberAnalysis.factorNumber(360u))
        assertEquals(factors(7, 7), NumberAnalysis.factorNumber(49u))
        assertEquals(factors(2, 3, 5, 7, 11, 13), NumberAnalysis.factorNumber(30030u))
        assertEquals(List(40) { 2.toULong() }, NumberAnalysis.factorNumber(1099511627776u))
    }

    @Test
    fun largeNumbersAreFactored() {
        assertEquals(factors(71, 839, 1471, 6857), NumberAnalysis.factorNumber(600851475143u))
        // Square of a prime, which is found only when the search goes all the way to the square root
        assertEquals(factors(999983, 999983), NumberAnalysis.factorNumber(999966000289u))
        assertEquals(factors(3, 5, 17, 257, 641, 65537, 6700417),
                NumberAnalysis.factorNumber(ULong.MAX_VALUE))
    }

    @Test
    fun romanNumeralsAreFormatted() {
        assertEquals("I", NumberAnalysis.formatRomanNumeral(1u))
        assertEquals("IV", NumberAnalysis.formatRomanNumeral(4u))
        assertEquals("IX", NumberAnalysis.formatRomanNumeral(9u))
        assertEquals("XIV", NumberAnalysis.formatRomanNumeral(14u))
        assertEquals("XL", NumberAnalysis.formatRomanNumeral(40u))
        assertEquals("XC", NumberAnalysis.formatRomanNumeral(90u))
        assertEquals("CD", NumberAnalysis.formatRomanNumeral(400u))
        assertEquals("MCMXCIV", NumberAnalysis.formatRomanNumeral(1994u))
        assertEquals("MMMCMXCIX", NumberAnalysis.formatRomanNumeral(3999u))
        assertEquals("MMMM", NumberAnalysis.formatRomanNumeral(4000u))
    }

    @Test
    fun romanNumeralsHaveLimits() {
        assertNull(NumberAnalysis.formatRomanNumeral(0u))
        assertEquals("M".repeat(100), NumberAnalysis.formatRomanNumeral(100000u))
        assertNull(NumberAnalysis.formatRomanNumeral(100001u))
    }

    @Test
    fun romanNumeralsAreParsed() {
        assertEquals(1994.toULong(), NumberAnalysis.parseRomanNumeral("MCMXCIV"))
        assertEquals(1994.toULong(), NumberAnalysis.parseRomanNumeral("mcmxciv"))
        assertEquals(4000.toULong(), NumberAnalysis.parseRomanNumeral("MMMM"))
    }

    @Test
    fun nonCanonicalRomanNumeralsAreRejected() {
        assertNull(NumberAnalysis.parseRomanNumeral(""))
        assertNull(NumberAnalysis.parseRomanNumeral("IIII"))
        assertNull(NumberAnalysis.parseRomanNumeral("IC"))
        assertNull(NumberAnalysis.parseRomanNumeral("VX"))
        assertNull(NumberAnalysis.parseRomanNumeral("XIVI"))
        assertNull(NumberAnalysis.parseRomanNumeral("ABC"))
        assertNull(NumberAnalysis.parseRomanNumeral("12"))
    }

    @Test
    fun romanNumeralsRoundTrip() {
        for (i in 1..5000) {
            val numeral = NumberAnalysis.formatRomanNumeral(i.toULong())!!
            assertEquals(numeral, i.toULong(), NumberAnalysis.parseRomanNumeral(numeral))
        }
    }

    @Test
    fun numbersAreParsedInGivenBase() {
        assertEquals(255.toULong(), NumberAnalysis.parseNumber("255", "base-10"))
        assertEquals(255.toULong(), NumberAnalysis.parseNumber("ff", "base-16"))
        assertEquals(255.toULong(), NumberAnalysis.parseNumber("FF", "base-16"))
        assertEquals(5.toULong(), NumberAnalysis.parseNumber("101", "base-2"))
        assertEquals(14.toULong(), NumberAnalysis.parseNumber("XIV", "roman numerals"))
        assertEquals(ULong.MAX_VALUE, NumberAnalysis.parseNumber("18446744073709551615", "base-10"))
    }

    @Test
    fun invalidNumbersAreNotParsed() {
        assertNull(NumberAnalysis.parseNumber("12", "base-2"))
        assertNull(NumberAnalysis.parseNumber("-1", "base-10"))
        assertNull(NumberAnalysis.parseNumber("1.5", "base-10"))
        assertNull(NumberAnalysis.parseNumber("18446744073709551616", "base-10"))
        assertNull(NumberAnalysis.parseNumber("XIV", "base-10"))
        assertNull(NumberAnalysis.parseNumber("14", "roman numerals"))
        assertNull(NumberAnalysis.parseNumber("5", "prime factors"))
    }

    @Test
    fun numbersAreFormattedInGivenBase() {
        assertEquals("255", NumberAnalysis.formatInBase(255u, 10))
        assertEquals("11111111", NumberAnalysis.formatInBase(255u, 2))
        assertEquals("FF", NumberAnalysis.formatInBase(255u, 16))
        assertEquals("FFFFFFFFFFFFFFFF", NumberAnalysis.formatInBase(ULong.MAX_VALUE, 16))
    }
}

File: pwa/test/number-analysis.test.ts
import { describe, expect, it } from 'vitest';
import {
  factorNumber, formatInBase, formatRomanNumeral, parseNumber, parseRomanNumeral, ULONG_MAX,
} from '../src/logic/number-analysis';

describe('factorNumber', () => {
  it('finds no factors below two', () => {
    expect(factorNumber(0n)).toEqual([]);
    expect(factorNumber(1n)).toEqual([]);
  });

  it('gives primes as their only factor', () => {
    expect(factorNumber(2n)).toEqual([2n]);
    expect(factorNumber(3n)).toEqual([3n]);
    expect(factorNumber(97n)).toEqual([97n]);
    expect(factorNumber(1000000007n)).toEqual([1000000007n]);
  });

  it('sorts and repeats factors', () => {
    expect(factorNumber(360n)).toEqual([2n, 2n, 2n, 3n, 3n, 5n]);
    expect(factorNumber(49n)).toEqual([7n, 7n]);
    expect(factorNumber(30030n)).toEqual([2n, 3n, 5n, 7n, 11n, 13n]);
    expect(factorNumber(1099511627776n)).toEqual(Array(40).fill(2n));
  });

  it('factors large numbers', () => {
    expect(factorNumber(600851475143n)).toEqual([71n, 839n, 1471n, 6857n]);
    // Square of a prime, which is found only when the search goes all the way to the square root
    expect(factorNumber(999966000289n)).toEqual([999983n, 999983n]);
    expect(factorNumber(ULONG_MAX)).toEqual([3n, 5n, 17n, 257n, 641n, 65537n, 6700417n]);
  });
});

describe('roman numerals', () => {
  it('are formatted', () => {
    expect(formatRomanNumeral(1n)).toBe('I');
    expect(formatRomanNumeral(4n)).toBe('IV');
    expect(formatRomanNumeral(9n)).toBe('IX');
    expect(formatRomanNumeral(14n)).toBe('XIV');
    expect(formatRomanNumeral(40n)).toBe('XL');
    expect(formatRomanNumeral(90n)).toBe('XC');
    expect(formatRomanNumeral(400n)).toBe('CD');
    expect(formatRomanNumeral(1994n)).toBe('MCMXCIV');
    expect(formatRomanNumeral(3999n)).toBe('MMMCMXCIX');
    expect(formatRomanNumeral(4000n)).toBe('MMMM');
  });

  it('have limits', () => {
    expect(formatRomanNumeral(0n)).toBeNull();
    expect(formatRomanNumeral(100000n)).toBe('M'.repeat(100));
    expect(formatRomanNumeral(100001n)).toBeNull();
  });

  it('are parsed', () => {
    expect(parseRomanNumeral('MCMXCIV')).toBe(1994n);
    expect(parseRomanNumeral('mcmxciv')).toBe(1994n);
    expect(parseRomanNumeral('MMMM')).toBe(4000n);
  });

  it('are rejected when not canonical', () => {
    expect(parseRomanNumeral('')).toBeNull();
    expect(parseRomanNumeral('IIII')).toBeNull();
    expect(parseRomanNumeral('IC')).toBeNull();
    expect(parseRomanNumeral('VX')).toBeNull();
    expect(parseRomanNumeral('XIVI')).toBeNull();
    expect(parseRomanNumeral('ABC')).toBeNull();
    expect(parseRomanNumeral('12')).toBeNull();
  });

  it('round trip', () => {
    for (let i = 1n; i <= 5000n; i++) {
      const numeral = formatRomanNumeral(i)!;
      expect(parseRomanNumeral(numeral), numeral).toBe(i);
    }
  });
});

describe('parseNumber', () => {
  it('parses numbers in given base', () => {
    expect(parseNumber('255', 'base-10')).toBe(255n);
    expect(parseNumber('ff', 'base-16')).toBe(255n);
    expect(parseNumber('FF', 'base-16')).toBe(255n);
    expect(parseNumber('101', 'base-2')).toBe(5n);
    expect(parseNumber('XIV', 'roman numerals')).toBe(14n);
    expect(parseNumber('18446744073709551615', 'base-10')).toBe(ULONG_MAX);
  });

  it('rejects invalid numbers', () => {
    expect(parseNumber('12', 'base-2')).toBeNull();
    expect(parseNumber('-1', 'base-10')).toBeNull();
    expect(parseNumber('1.5', 'base-10')).toBeNull();
    expect(parseNumber('18446744073709551616', 'base-10')).toBeNull();
    expect(parseNumber('XIV', 'base-10')).toBeNull();
    expect(parseNumber('14', 'roman numerals')).toBeNull();
    expect(parseNumber('5', 'prime factors')).toBeNull();
  });
});

describe('formatInBase', () => {
  it('formats numbers in given base', () => {
    expect(formatInBase(255n, 10)).toBe('255');
    expect(formatInBase(255n, 2)).toBe('11111111');
    expect(formatInBase(255n, 16)).toBe('FF');
    expect(formatInBase(ULONG_MAX, 16)).toBe('FFFFFFFFFFFFFFFF');
  });
});
